File: sdk/typescript/src/scoped_jwt.test.ts
import { describe, it, expect, beforeAll } from "vitest";
// jose v6：无 KeyLike 导出，密钥类型用 CryptoKey（importSPKI/generateKeyPair 的返回类型）。
import { SignJWT, exportSPKI, generateKeyPair, type CryptoKey } from "jose";
import { ScopedJWTVerifier } from "./scoped_jwt";
import { TokenExpiredError, TokenAudienceMismatchError, TokenInvalidError } from "./errors";

let priv: CryptoKey;
let spki: string;
const KID = "test-kid";

beforeAll(async () => {
  const kp = await generateKeyPair("RS256");
  priv = kp.privateKey;
  spki = await exportSPKI(kp.publicKey);
});

async function sign(claims: Record<string, unknown>, opts: { aud: string; expSec?: number }) {
  const now = Math.floor(Date.now() / 1000);
  return await new SignJWT(claims)
    .setProtectedHeader({ alg: "RS256", kid: KID })
    .setIssuedAt(now)
    .setExpirationTime(now + (opts.expSec ?? 300))
    .setAudience(opts.aud)
    .sign(priv);
}

describe("ScopedJWTVerifier.verify（kx_ claims）", () => {
  it("happy：映射 kx_caller_id 等", async () => {
    const v = new ScopedJWTVerifier();
    await v.setStaticKey(KID, spki);
    const token = await sign(
      { sub: "7", kx_caller_id: "app-7", kx_caller_kind: "app", kx_chain_id: "chn_1", kx_request_id: "req-1" },
      { aud: "ks-mcp-x.foo" },
    );
    const claims = await v.verify(token, "ks-mcp-x.foo");
    expect(claims.userId).toBe("7");
    expect(claims.canonicalName).toBe("ks-mcp-x.foo");
    expect(claims.callerId).toBe("app-7");
    expect(claims.chainId).toBe("chn_1");
  });

  it("过期 → TokenExpired", async () => {
    const v = new ScopedJWTVerifier();
    await v.setStaticKey(KID, spki);
    const token = await sign({ sub: "7" }, { aud: "ks-mcp-x.foo", expSec: -10 });
    await expect(v.verify(token, "ks-mcp-x.foo")).rejects.toBeInstanceOf(TokenExpiredError);
  });

  it("aud 不符 → TokenAudienceMismatch", async () => {
    const v = new ScopedJWTVerifier();
    await v.setStaticKey(KID, spki);
    const token = await sign({ sub: "7" }, { aud: "ks-mcp-x.other" });
    await expect(v.verify(token, "ks-mcp-x.foo")).rejects.toBeInstanceOf(TokenAudienceMismatchError);
  });

  it("未知 kid → TokenInvalid", async () => {
    const v = new ScopedJWTVerifier();
    const token = await sign({ sub: "7" }, { aud: "ks-mcp-x.foo" });
    await expect(v.verify(token, "ks-mcp-x.foo")).rejects.toBeInstanceOf(TokenInvalidError);
  });
});

File: sdk/go/ksapp/scoped_jwt_test.go
package ksapp

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"errors"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

func makeRSAKeypair(t *testing.T) (*rsa.PrivateKey, string) {
	t.Helper()
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatal(err)
	}
	pubBytes, _ := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	pubPEM := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: pubBytes})
	return priv, string(pubPEM)
}

func signRS256(t *testing.T, priv *rsa.PrivateKey, claims jwt.MapClaims) string {
	t.Helper()
	tok := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	tok.Header["kid"] = "test-kid"
	s, err := tok.SignedString(priv)
	if err != nil {
		t.Fatal(err)
	}
	return s
}

func TestScopedJWTVerifierVerifyOK(t *testing.T) {
	priv, pubPEM := makeRSAKeypair(t)
	now := time.Now().Unix()
	token := signRS256(t, priv, jwt.MapClaims{
		"iss":            "keystone",
		"aud":            "ks.x.foo",
		"sub":            "u-1",
		"iat":            now,
		"exp":            now + 60,
		"kx_caller_id":   "ks-mcp-writer",
		"kx_caller_kind": "app",
		"kx_chain_id":    "chain-1",
		"kx_request_id":  "req-1",
	})
	v := NewScopedJWTVerifier("")
	if err := v.SetStaticKey("test-kid", pubPEM); err != nil {
		t.Fatal(err)
	}
	claims, err := v.Verify(token, "ks.x.foo")
	if err != nil {
		t.Fatal(err)
	}
	if claims.UserID != "u-1" {
		t.Fatalf("UserID = %q", claims.UserID)
	}
	if claims.CanonicalName != "ks.x.foo" {
		t.Fatalf("CanonicalName = %q", claims.CanonicalName)
	}
	if claims.CallerID != "ks-mcp-writer" {
		t.Fatalf("CallerID = %q", claims.CallerID)
	}
	if claims.CallerKind != "app" {
		t.Fatalf("CallerKind = %q", claims.CallerKind)
	}
	if claims.ChainID != "chain-1" {
		t.Fatalf("ChainID = %q", claims.ChainID)
	}
	if claims.RequestID != "req-1" {
		t.Fatalf("RequestID = %q", claims.RequestID)
	}
}

func TestScopedJWTVerifierWrongAud(t *testing.T) {
	priv, pubPEM := makeRSAKeypair(t)
	now := time.Now().Unix()
	token := signRS256(t, priv, jwt.MapClaims{
		"iss": "keystone", "aud": "ks.x.OTHER",
		"sub": "u", "iat": now, "exp": now + 60,
	})
	v := NewScopedJWTVerifier("")
	if err := v.SetStaticKey("test-kid", pubPEM); err != nil {
		t.Fatal(err)
	}
	_, err := v.Verify(token, "ks.x.foo")
	if !errors.Is(err, ErrTokenAudienceMismatch) {
		t.Fatalf("expected aud mismatch, got %v", err)
	}
}

func TestScopedJWTVerifierExpired(t *testing.T) {
	priv, pubPEM := makeRSAKeypair(t)
	now := time.Now().Unix()
	token := signRS256(t, priv, jwt.MapClaims{
		"iss": "keystone", "aud": "ks.x.foo",
		"sub": "u", "iat": now - 120, "exp": now - 60,
	})
	v := NewScopedJWTVerifier("")
	if err := v.SetStaticKey("test-kid", pubPEM); err != nil {
		t.Fatal(err)
	}
	_, err := v.Verify(token, "ks.x.foo")
	if !errors.Is(err, ErrTokenExpired) {
		t.Fatalf("expected token expired, got %v", err)
	}
}

func TestScopedJWTVerifierInvalidSignature(t *testing.T) {
	priv, _ := makeRSAKeypair(t)
	_, otherPub := makeRSAKeypair(t)
	now := time.Now().Unix()
	token := signRS256(t, priv, jwt.MapClaims{
		"iss": "keystone", "aud": "ks.x.foo",
		"sub": "u", "iat": now, "exp": now + 60,
	})
	v := NewScopedJWTVerifier("")
	if err := v.SetStaticKey("test-kid", otherPub); err != nil {
		t.Fatal(err)
	}
	_, err := v.Verify(token, "ks.x.foo")
	if !errors.Is(err, ErrTokenInvalid) {
		t.Fatalf("expected token invalid, got %v", err)
	}
}

func TestScopedJWTVerifierUnknownKid(t *testing.T) {
	priv, _ := makeRSAKeypair(t)
	now := time.Now().Unix()
	token := signRS256(t, priv, jwt.MapClaims{
		"iss": "keystone", "aud": "ks.x.foo",
		"sub": "u", "iat": now, "exp": now + 60,
	})
	v := NewScopedJWTVerifier("")
	_, err := v.Verify(token, "ks.x.foo")
	if !errors.Is(err, ErrTokenInvalid) {
		t.Fatalf("expected token invalid (unknown kid), got %v", err)
	}
}
